Sources/SwiftProtobuf/BinaryDecoder.swift

```diff
@@ -18,1477 +18,1505 @@
+      private var fieldNumber: Int = 0
+      // Collection of extension fields for this decode
+      private var extensions: ExtensionMap?
+      // The current group number. See decodeFullGroup(group:fieldNumber:) for how
+      // this is used.
+      private var groupFieldNumber: Int?
+      // The options for decoding.
+      private var options: BinaryDecodingOptions
+
+      private var recursionBudget: Int
+
+      // Collects the unknown data found while decoding a message.
+      private var unknownData: Data?
+      // Custom data to use as the unknown data while parsing a field. Used only by
+      // packed repeated enums; see below
+      private var unknownOverride: Data?
+
+      private var complete: Bool { return available == 0 }
+
+      internal init(
+            forReadingFrom pointer: UnsafePointer<UInt8>, count: Int,
+            options: BinaryDecodingOptions, extensions: ExtensionMap? = nil
+      ) {
+            // Assuming baseAddress is not nil.
+            p = pointer
+            available = count
+            fieldStartP = p
+            self.extensions = extensions
+            self.options = options
+            recursionBudget = options.messageDepthLimit
+      }
+
+      internal init(
+            forReadingFrom pointer: UnsafePointer<UInt8>, count: Int,
+            parent: BinaryDecoder
+      ) {
+            self.init(
+                  forReadingFrom: pointer, count: count,
+                  options: parent.options, extensions: parent.extensions)
+            recursionBudget = parent.recursionBudget
+      }
+
+      private mutating func incrementRecursionDepth() throws {
+            recursionBudget -= 1
+            if recursionBudget < 0 {
+                  throw BinaryDecodingError.messageDepthLimit
             }
-        }
-
-        // Quit if end of input
-        if available == 0 {
-            return nil
-        }
-
-        // Get the next field number
-        fieldStartP = p
-        fieldEndP = nil
-        let start = p
-        let c0 = start[0]
-        if let wireFormat = WireFormat(rawValue: c0 & 7) {
-            fieldWireFormat = wireFormat
-        } else {
-            throw BinaryDecodingError.malformedProtobuf
-        }
-        if (c0 & 0x80) == 0 {
-            p += 1
-            available -= 1
-            fieldNumber = Int(c0) >> 3
-        } else {
-            fieldNumber = Int(c0 & 0x7f) >> 3
-            if available < 2 {
-                throw BinaryDecodingError.malformedProtobuf
+      }
+
+      private mutating func decrementRecursionDepth() {
+            recursionBudget += 1
+            // This should never happen, if it does, something is probably corrupting memory, and
+            // simply throwing doesn't make much sense.
+            if recursionBudget > options.messageDepthLimit {
+                  fatalError(
+                        "Somehow BinaryDecoding unwound more objects than it started"
+                  )
```

Comment:
The string exceeds 80 columns; it should be rewritten so it fits.

```swift
print("""
  no \
  newline
  """)
```

Reply:
We could consider adding a `--swift-version` flag to the formatter that would let us conditionally enable certain transformations only if the target version supports it.

I'm not sure what the right default would be though if it's not specified (the current toolchain version?) and how we would query that cleanly in the context of the formatter.

Reply:
Just for comments currently.

That being said, there's no good reason that we only do it for comments, and strings aren't a special case either; depending on the indentation, line length, and tokens involved, it's possible to end up with degenerate cases where there's no good way to break a token without it exceeding the limit, so we should unify our handling of these.